File: app/src/lib/utils/calendar.ts
/**
 * Calendar Integration Utilities for Night Owls Shift Bookings
 * Supports .ics generation, calendar URLs, and popular calendar app integration
 *
 * TODO: Extract shared calendar generation logic to avoid duplication with backend
 * This frontend implementation should eventually use a shared library for consistency
 */

import type { UserBooking } from '$lib/services/api/user';
import { formatDateTime } from '$lib/utils/datetime';

// === TYPES ===

export interface CalendarEvent {
	title: string;
	description: string;
	startTime: Date;
	endTime: Date;
	location?: string;
	organizer?: {
		name: string;
		email: string;
	};
	attendees?: Array<{
		name: string;
		email?: string;
	}>;
	reminder?: {
		minutes: number;
	};
	uid: string;
}

export interface CalendarProvider {
	name: string;
	url: string;
	icon: string;
}

// === CONSTANTS ===

const CALENDAR_PROVIDERS = {
	google: {
		name: 'Google Calendar',
		icon: '📅',
		baseUrl: 'https://calendar.google.com/calendar/render'
	},
	outlook: {
		name: 'Outlook',
		icon: '📘',
		baseUrl: 'https://outlook.live.com/calendar/0/deeplink/compose'
	},
	apple: {
		name: 'Apple Calendar',
		icon: '🍎',
		baseUrl: 'data:text/calendar;charset=utf8'
	},
	yahoo: {
		name: 'Yahoo Calendar',
		icon: '🟣',
		baseUrl: 'https://calendar.yahoo.com'
	}
} as const;

// === CORE CALENDAR FUNCTIONS ===

/**
 * Convert UserBooking to CalendarEvent
 */
export function bookingToCalendarEvent(booking: UserBooking): CalendarEvent {
	const startTime = new Date(booking.shift_start);
	const endTime = new Date(booking.shift_end);

	// Create detailed description
	let description = `Night Owls Community Watch - ${booking.schedule_name}\n\n`;
	description += `⏰ Time: ${formatDateTime(booking.shift_start)} - ${formatDateTime(booking.shift_end)}\n`;

	if (booking.buddy_name) {
		description += `👥 Buddy: ${booking.buddy_name}\n`;
	}

	description += `\n📱 Check in on the Night Owls app when your shift starts`;
	description += `\n🚨 Report any incidents through the app`;
	description += `\n\n🔗 Night Owls App: ${typeof window !== 'undefined' ? window.location.origin : 'https://your-domain.com'}`;

	// Create attendees list
	const attendees: CalendarEvent['attendees'] = [];
	if (booking.buddy_name) {
		attendees.push({ name: booking.buddy_name });
	}

	return {
		title: `Night Owls Shift - ${booking.schedule_name}`,
		description,
		startTime,
		endTime,
		location: 'Mount Moreland Community Watch Area',
		organizer: {
			name: 'Night Owls Scheduler',
			email: 'noreply@mm.nightowls.app'
		},
		attendees,
		reminder: {
			minutes: 60 // 1 hour before shift
		},
		uid: `nightowls-shift-${booking.booking_id}@mm.nightowls.app`
	};
}

/**
 * Generate .ics file content for a single event
 */
export function generateICSEvent(event: CalendarEvent): string {
	const formatICSDate = (date: Date): string => {
		return date
			.toISOString()
			.replace(/[:-]/g, '')
			.replace(/\.\d{3}/, '');
	};

	const escapeICSText = (text: string): string => {
		return text
			.replace(/\\/g, '\\\\')
			.replace(/;/g, '\\;')
			.replace(/,/g, '\\,')
			.replace(/\n/g, '\\n');
	};

	const now = new Date();
	const icsContent = [
		'BEGIN:VCALENDAR',
		'VERSION:2.0',
		'PRODID:-//Night Owls//Shift Calendar//EN',
		'CALSCALE:GREGORIAN',
		'METHOD:PUBLISH',
		'BEGIN:VEVENT',
		`UID:${event.uid}`,
		`DTSTAMP:${formatICSDate(now)}`,
		`DTSTART:${formatICSDate(event.startTime)}`,
		`DTEND:${formatICSDate(event.endTime)}`,
		`SUMMARY:${escapeICSText(event.title)}`,
		`DESCRIPTION:${escapeICSText(event.description)}`,
		...(event.location ? [`LOCATION:${escapeICSText(event.location)}`] : []),
		...(event.organizer
			? [`ORGANIZER;CN=${escapeICSText(event.organizer.name)}:mailto:${event.organizer.email}`]
			: []),
		...(event.attendees?.map(
			(attendee) =>
				`ATTENDEE;CN=${escapeICSText(attendee.name)}${attendee.email ? `:mailto:${attendee.email}` : ''}`
		) || []),
		...(event.reminder
			? [
					'BEGIN:VALARM',
					'TRIGGER:-PT' + event.reminder.minutes + 'M',
					'ACTION:DISPLAY',
					`DESCRIPTION:${escapeICSText(event.title)} starts in ${event.reminder.minutes} minutes`,
					'END:VALARM'
				]
			: []),
		'END:VEVENT',
		'END:VCALENDAR'
	];

	return icsContent.join('\r\n');
}

/**
 * Generate .ics file content for multiple events
 */
export function generateICSCalendar(
	events: CalendarEvent[],
	calendarName = 'Night Owls Shifts'
): string {
	const formatICSDate = (date: Date): string => {
		return date
			.toISOString()
			.replace(/[:-]/g, '')
			.replace(/\.\d{3}/, '');
	};

	const escapeICSText = (text: string): string => {
		return text
			.replace(/\\/g, '\\\\')
			.replace(/;/g, '\\;')
			.replace(/,/g, '\\,')
			.replace(/\n/g, '\\n');
	};

	const now = new Date();

	const icsContent = [
		'BEGIN:VCALENDAR',
		'VERSION:2.0',
		'PRODID:-//Night Owls//Shift Calendar//EN',
		'CALSCALE:GREGORIAN',
		'METHOD:PUBLISH',
		`X-WR-CALNAME:${escapeICSText(calendarName)}`,
		'X-WR-CALDESC:Mount Moreland Night Owls Community Watch Shifts',
		'X-WR-TIMEZONE:Africa/Johannesburg'
	];

	// Add each event
	for (const event of events) {
		icsContent.push(
			'BEGIN:VEVENT',
			`UID:${event.uid}`,
			`DTSTAMP:${formatICSDate(now)}`,
			`DTSTART:${formatICSDate(event.startTime)}`,
			`DTEND:${formatICSDate(event.endTime)}`,
			`SUMMARY:${escapeICSText(event.title)}`,
			`DESCRIPTION:${escapeICSText(event.description)}`,
			...(event.location ? [`LOCATION:${escapeICSText(event.location)}`] : []),
			...(event.organizer
				? [`ORGANIZER;CN=${escapeICSText(event.organizer.name)}:mailto:${event.organizer.email}`]
				: []),
			...(event.attendees?.map(
				(attendee) =>
					`ATTENDEE;CN=${escapeICSText(attendee.name)}${attendee.email ? `:mailto:${attendee.email}` : ''}`
			) || []),
			...(event.reminder
				? [
						'BEGIN:VALARM',
						'TRIGGER:-PT' + event.reminder.minutes + 'M',
						'ACTION:DISPLAY',
						`DESCRIPTION:${escapeICSText(event.title)} starts in ${event.reminder.minutes} minutes`,
						'END:VALARM'
					]
				: []),
			'END:VEVENT'
		);
	}

	icsContent.push('END:VCALENDAR');
	return icsContent.join('\r\n');
}

// === CALENDAR PROVIDER URLS ===

/**
 * Generate Google Calendar URL
 */
export function generateGoogleCalendarURL(event: CalendarEvent): string {
	const formatGoogleDate = (date: Date): string => {
		return date
			.toISOString()
			.replace(/[:-]/g, '')
			.replace(/\.\d{3}/, '');
	};

	const params = new URLSearchParams({
		action: 'TEMPLATE',
		text: event.title,
		dates: `${formatGoogleDate(event.startTime)}/${formatGoogleDate(event.endTime)}`,
		details: event.description,
		...(event.location && { location: event.location })
	});

	return `${CALENDAR_PROVIDERS.google.baseUrl}?${params.toString()}`;
}

/**
 * Generate Outlook Calendar URL
 */
export function generateOutlookURL(event: CalendarEvent): string {
	const params = new URLSearchParams({
		subject: event.title,
		startdt: event.startTime.toISOString(),
		enddt: event.endTime.toISOString(),
		body: event.description,
		...(event.location && { location: event.location })
	});

	return `${CALENDAR_PROVIDERS.outlook.baseUrl}?${params.toString()}`;
}

/**
 * Generate Yahoo Calendar URL
 */
export function generateYahooURL(event: CalendarEvent): string {
	const formatYahooDate = (date: Date): string => {
		return date
			.toISOString()
			.replace(/[:-]/g, '')
			.replace(/\.\d{3}/, '');
	};

	const params = new URLSearchParams({
		v: '60',
		title: event.title,
		st: formatYahooDate(event.startTime),
		et: formatYahooDate(event.endTime),
		desc: event.description,
		...(event.location && { in_loc: event.location })
	});

	return `${CALENDAR_PROVIDERS.yahoo.baseUrl}?${params.toString()}`;
}

// === FILE DOWNLOAD UTILITIES ===

/**
 * Download .ics file for a single booking
 */
export function downloadShiftICS(booking: UserBooking): void {
	const event = bookingToCalendarEvent(booking);
	const icsContent = generateICSEvent(event);
	const fileName = `night-owls-shift-${booking.booking_id}.ics`;

	downloadICSFile(icsContent, fileName);
}

/**
 * Download .ics file for multiple bookings
 */
export function downloadAllShiftsICS(bookings: UserBooking[]): void {
	const events = bookings.map(bookingToCalendarEvent);
	const icsContent = generateICSCalendar(events, 'Night Owls Shifts');
	const fileName = `night-owls-all-shifts-${new Date().toISOString().split('T')[0]}.ics`;

	downloadICSFile(icsContent, fileName);
}

/**
 * Utility to trigger file download
 */
function downloadICSFile(content: string, fileName: string): void {
	const blob = new Blob([content], { type: 'text/calendar;charset=utf-8' });
	const url = URL.createObjectURL(blob);

	const link = document.createElement('a');
	link.href = url;
	link.download = fileName;
	link.style.display = 'none';

	document.body.appendChild(link);
	link.click();
	document.body.removeChild(link);

	URL.revokeObjectURL(url);
}

// === CALENDAR PROVIDER INTEGRATION ===

/**
 * Get all supported calendar providers with their URLs for an event
 */
export function getCalendarProviders(event: CalendarEvent): CalendarProvider[] {
	return [
		{
			name: CALENDAR_PROVIDERS.google.name,
			url: generateGoogleCalendarURL(event),
			icon: CALENDAR_PROVIDERS.google.icon
		},
		{
			name: CALENDAR_PROVIDERS.outlook.name,
			url: generateOutlookURL(event),
			icon: CALENDAR_PROVIDERS.outlook.icon
		},
		{
			name: CALENDAR_PROVIDERS.yahoo.name,
			url: generateYahooURL(event),
			icon: CALENDAR_PROVIDERS.yahoo.icon
		}
	];
}

/**
 * Open calendar provider in new window/tab
 */
export function openCalendarProvider(url: string): void {
	window.open(url, '_blank', 'noopener,noreferrer');
}

// === UTILITY FUNCTIONS ===

/**
 * Check if a booking is in the future (can be added to calendar)
 */
export function canAddToCalendar(booking: UserBooking): boolean {
	const shiftStart = new Date(booking.shift_start);
	const now = new Date();
	return shiftStart > now;
}

/**
 * Get calendar integration summary for bookings
 */
export function getCalendarSummary(bookings: UserBooking[]): {
	total: number;
	upcoming: number;
	thisWeek: number;
	thisMonth: number;
} {
	const now = new Date();
	const oneWeek = new Date(now.getTime() + 7 * 24 * 60 * 60 * 1000);
	const oneMonth = new Date(now.getTime() + 30 * 24 * 60 * 60 * 1000);

	const upcoming = bookings.filter((booking) => {
		const shiftStart = new Date(booking.shift_start);
		return shiftStart > now;
	});

	return {
		total: bookings.length,
		upcoming: upcoming.length,
		thisWeek: upcoming.filter((booking) => {
			const shiftStart = new Date(booking.shift_start);
			return shiftStart <= oneWeek;
		}).length,
		thisMonth: upcoming.filter((booking) => {
			const shiftStart = new Date(booking.shift_start);
			return shiftStart <= oneMonth;
		}).length
	};
}

File: internal/utils/calendar.go
// Package utils provides utility functions for calendar generation
package utils

import (
	"fmt"
	"strings"
	"time"

	db "night-owls-go/internal/db/sqlc_generated"
)

// CalendarEvent represents a calendar event
type CalendarEvent struct {
	Title       string
	Description string
	StartTime   time.Time
	EndTime     time.Time
	Location    string
	UID         string
	Organizer   CalendarContact
	Attendees   []CalendarContact
	ReminderMin int // Minutes before event
}

// CalendarContact represents a contact in calendar
type CalendarContact struct {
	Name  string
	Email string
}

// ICSData contains calendar file information
type ICSData struct {
	Filename string
	Content  string
	MIME     string
}

// BookingToCalendarEvent converts a booking to a calendar event
func BookingToCalendarEvent(booking db.Booking, scheduleName string) CalendarEvent {
	title := fmt.Sprintf("Night Owls Shift - %s", scheduleName)
	
	// Build detailed description
	var descBuilder strings.Builder
	descBuilder.WriteString(fmt.Sprintf("Night Owls Community Watch - %s\\n\\n", scheduleName))
	descBuilder.WriteString(fmt.Sprintf("⏰ Time: %s - %s\\n", 
		booking.ShiftStart.Format("15:04"), 
		booking.ShiftEnd.Format("15:04")))
	
	if booking.BuddyName.Valid && booking.BuddyName.String != "" {
		descBuilder.WriteString(fmt.Sprintf("👥 Buddy: %s\\n", booking.BuddyName.String))
	}
	
	descBuilder.WriteString("\\n📱 Check in on the Night Owls app when your shift starts")
	descBuilder.WriteString("\\n🚨 Report any incidents through the app")
	descBuilder.WriteString("\\n\\n🔗 Night Owls App: https://mm.nightowls.app")
	
	// Create attendees list
	var attendees []CalendarContact
	if booking.BuddyName.Valid && booking.BuddyName.String != "" {
		attendees = append(attendees, CalendarContact{
			Name:  booking.BuddyName.String,
			Email: "",
		})
	}
	
	return CalendarEvent{
		Title:       title,
		Description: descBuilder.String(),
		StartTime:   booking.ShiftStart,
		EndTime:     booking.ShiftEnd,
		Location:    "Mount Moreland Community Watch Area",
		UID:         fmt.Sprintf("nightowls-shift-%d@mm.nightowls.app", booking.BookingID),
		Organizer: CalendarContact{
			Name:  "Night Owls Scheduler",
			Email: "noreply@mm.nightowls.app",
		},
		Attendees:   attendees,
		ReminderMin: 60, // 1 hour before shift
	}
}

// GenerateICS creates an .ics file content from a calendar event
func GenerateICS(event CalendarEvent) string {
	now := time.Now().UTC()
	
	var icsBuilder strings.Builder
	
	// ICS Header
	icsBuilder.WriteString("BEGIN:VCALENDAR\r\n")
	icsBuilder.WriteString("VERSION:2.0\r\n")
	icsBuilder.WriteString("PRODID:-//Night Owls//Shift Calendar//EN\r\n")
	icsBuilder.WriteString("CALSCALE:GREGORIAN\r\n")
	icsBuilder.WriteString("METHOD:PUBLISH\r\n")
	
	// Event
	icsBuilder.WriteString("BEGIN:VEVENT\r\n")
	icsBuilder.WriteString(fmt.Sprintf("UID:%s\r\n", event.UID))
	icsBuilder.WriteString(fmt.Sprintf("DTSTAMP:%s\r\n", formatICSDate(now)))
	icsBuilder.WriteString(fmt.Sprintf("DTSTART:%s\r\n", formatICSDate(event.StartTime)))
	icsBuilder.WriteString(fmt.Sprintf("DTEND:%s\r\n", formatICSDate(event.EndTime)))
	icsBuilder.WriteString(fmt.Sprintf("SUMMARY:%s\r\n", escapeICSText(event.Title)))
	icsBuilder.WriteString(fmt.Sprintf("DESCRIPTION:%s\r\n", escapeICSText(event.Description)))
	
	if event.Location != "" {
		icsBuilder.WriteString(fmt.Sprintf("LOCATION:%s\r\n", escapeICSText(event.Location)))
	}
	
	// Organizer
	if event.Organizer.Email != "" {
		icsBuilder.WriteString(fmt.Sprintf("ORGANIZER;CN=%s:mailto:%s\r\n", 
			escapeICSText(event.Organizer.Name), event.Organizer.Email))
	}
	
	// Attendees
	for _, attendee := range event.Attendees {
		if attendee.Email != "" {
			icsBuilder.WriteString(fmt.Sprintf("ATTENDEE;CN=%s:mailto:%s\r\n", 
				escapeICSText(attendee.Name), attendee.Email))
		} else {
			icsBuilder.WriteString(fmt.Sprintf("ATTENDEE;CN=%s\r\n", 
				escapeICSText(attendee.Name)))
		}
	}
	
	// Reminder
	if event.ReminderMin > 0 {
		icsBuilder.WriteString("BEGIN:VALARM\r\n")
		icsBuilder.WriteString(fmt.Sprintf("TRIGGER:-PT%dM\r\n", event.ReminderMin))
		icsBuilder.WriteString("ACTION:DISPLAY\r\n")
		icsBuilder.WriteString(fmt.Sprintf("DESCRIPTION:%s starts in %d minutes\r\n", 
			escapeICSText(event.Title), event.ReminderMin))
		icsBuilder.WriteString("END:VALARM\r\n")
	}
	
	icsBuilder.WriteString("END:VEVENT\r\n")
	icsBuilder.WriteString("END:VCALENDAR\r\n")
	
	return icsBuilder.String()
}

// GenerateBookingICS creates an ICS file for a single booking
func GenerateBookingICS(booking db.Booking, scheduleName string) ICSData {
	event := BookingToCalendarEvent(booking, scheduleName)
	content := GenerateICS(event)
	filename := fmt.Sprintf("night-owls-shift-%d.ics", booking.BookingID)
	
	return ICSData{
		Filename: filename,
		Content:  content,
		MIME:     "text/calendar; charset=utf-8",
	}
}

// GenerateUserCalendarFeed creates a multi-event ICS feed for WebCal subscription
func GenerateUserCalendarFeed(bookings []db.ListBookingsByUserIDWithScheduleRow, userID int64) ICSData {
	if len(bookings) == 0 {
		// Empty calendar
		content := generateEmptyCalendar(userID)
		return ICSData{
			Filename: fmt.Sprintf("night-owls-calendar-%d.ics", userID),
			Content:  content,
			MIME:     "text/calendar; charset=utf-8",
		}
	}
	
	// Generate header
	var sb strings.Builder
	sb.WriteString("BEGIN:VCALENDAR\r\n")
	sb.WriteString("VERSION:2.0\r\n")
	sb.WriteString("PRODID:-//Night Owls Community Watch//Calendar Feed//EN\r\n")
	sb.WriteString("CALSCALE:GREGORIAN\r\n")
	sb.WriteString("METHOD:PUBLISH\r\n")
	sb.WriteString("X-WR-CALNAME:Night Owls Shifts\r\n")
	sb.WriteString("X-WR-CALDESC:Your upcoming Night Owls Community Watch shifts\r\n")
	sb.WriteString("X-WR-TIMEZONE:Africa/Johannesburg\r\n")
	
	// Add each booking as an event
	for _, booking := range bookings {
		// Only include future shifts in WebCal feed
		if booking.ShiftStart.After(time.Now()) {
			event := bookingRowToCalendarEvent(booking)
			eventContent := generateVEvent(event)
			sb.WriteString(eventContent)
		}
	}
	
	sb.WriteString("END:VCALENDAR\r\n")
	
	return ICSData{
		Filename: fmt.Sprintf("night-owls-calendar-%d.ics", userID),
		Content:  sb.String(),
		MIME:     "text/calendar; charset=utf-8",
	}
}

// bookingRowToCalendarEvent converts a booking row to a calendar event
func bookingRowToCalendarEvent(booking db.ListBookingsByUserIDWithScheduleRow) CalendarEvent {
	// Calculate shift end time (assuming 2-hour shifts if not specified)
	shiftEnd := booking.ShiftEnd
	if shiftEnd.IsZero() {
		shiftEnd = booking.ShiftStart.Add(2 * time.Hour)
	}
	
	// Build description
	var description strings.Builder
	description.WriteString(fmt.Sprintf("Night Owls Community Watch shift for %s\\n\\n", booking.ScheduleName))
	description.WriteString(fmt.Sprintf("📅 Date: %s\\n", booking.ShiftStart.Format("Monday, 2 January 2006")))
	description.WriteString(fmt.Sprintf("🕒 Time: %s - %s\\n", 
		booking.ShiftStart.Format("15:04"), 
		shiftEnd.Format("15:04")))
	
	if booking.BuddyName.Valid && booking.BuddyName.String != "" {
		description.WriteString(fmt.Sprintf("👥 Buddy: %s\\n", escapeICSText(booking.BuddyName.String)))
	}
	
	description.WriteString("\\n📱 Check in through the Night Owls app when your shift starts.")
	description.WriteString("\\n🔗 App: https://mm.nightowls.app")
	
	// Generate unique UID for this booking
	uid := fmt.Sprintf("night-owls-booking-%d@mm.nightowls.app", booking.BookingID)
	
	return CalendarEvent{
		Title:       fmt.Sprintf("Night Owls Shift - %s", booking.ScheduleName),
		Description: description.String(),
		StartTime:   booking.ShiftStart,
		EndTime:     shiftEnd,
		Location:    "Mount Moreland Community Watch Area",
		UID:         uid,
		Organizer: CalendarContact{
			Name:  "Night Owls Scheduler",
			Email: "noreply@mm.nightowls.app",
		},
		Attendees:   []CalendarContact{},
		ReminderMin: 60, // 1 hour before
	}
}

// generateEmptyCalendar creates an empty calendar for users with no shifts
func generateEmptyCalendar(userID int64) string {
	var sb strings.Builder
	sb.WriteString("BEGIN:VCALENDAR\r\n")
	sb.WriteString("VERSION:2.0\r\n")
	sb.WriteString("PRODID:-//Night Owls Community Watch//Calendar Feed//EN\r\n")
	sb.WriteString("CALSCALE:GREGORIAN\r\n")
	sb.WriteString("METHOD:PUBLISH\r\n")
	sb.WriteString("X-WR-CALNAME:Night Owls Shifts\r\n")
	sb.WriteString("X-WR-CALDESC:Your upcoming Night Owls Community Watch shifts\r\n")
	sb.WriteString("X-WR-TIMEZONE:Africa/Johannesburg\r\n")
	sb.WriteString("END:VCALENDAR\r\n")
	
	return sb.String()
}

// generateVEvent creates a VEVENT block for a calendar event
func generateVEvent(event CalendarEvent) string {
	var sb strings.Builder
	
	sb.WriteString("BEGIN:VEVENT\r\n")
	sb.WriteString(fmt.Sprintf("UID:%s\r\n", event.UID))
	sb.WriteString(fmt.Sprintf("DTSTAMP:%s\r\n", time.Now().UTC().Format("20060102T150405Z")))
	sb.WriteString(fmt.Sprintf("DTSTART:%s\r\n", event.StartTime.UTC().Format("20060102T150405Z")))
	sb.WriteString(fmt.Sprintf("DTEND:%s\r\n", event.EndTime.UTC().Format("20060102T150405Z")))
	sb.WriteString(fmt.Sprintf("SUMMARY:%s\r\n", escapeICSText(event.Title)))
	sb.WriteString(fmt.Sprintf("DESCRIPTION:%s\r\n", event.Description))
	sb.WriteString(fmt.Sprintf("LOCATION:%s\r\n", escapeICSText(event.Location)))
	
	if event.Organizer.Email != "" {
		sb.WriteString(fmt.Sprintf("ORGANIZER;CN=%s:MAILTO:%s\r\n", 
			escapeICSText(event.Organizer.Name), event.Organizer.Email))
	}
	
	// Add reminder alarm
	if event.ReminderMin > 0 {
		sb.WriteString("BEGIN:VALARM\r\n")
		sb.WriteString(fmt.Sprintf("TRIGGER:-PT%dM\r\n", event.ReminderMin))
		sb.WriteString("ACTION:DISPLAY\r\n")
		sb.WriteString(fmt.Sprintf("DESCRIPTION:Night Owls shift starts in %d minutes\r\n", event.ReminderMin))
		sb.WriteString("END:VALARM\r\n")
	}
	
	sb.WriteString("END:VEVENT\r\n")
	
	return sb.String()
}

// Helper functions

// formatICSDate formats a time for ICS format (YYYYMMDDTHHMMSSZ)
func formatICSDate(t time.Time) string {
	return t.UTC().Format("20060102T150405Z")
}

// escapeICSText escapes special characters for ICS format
func escapeICSText(text string) string {
	text = strings.ReplaceAll(text, "\\", "\\\\")
	text = strings.ReplaceAll(text, ";", "\\;")
	text = strings.ReplaceAll(text, ",", "\\,")
	text = strings.ReplaceAll(text, "\n", "\\n")
	return text
} 